dbms/src/Flash/tests/gtest_spill_aggregation.cpp

```diff
@@ -0,0 +1,99 @@
+{
+public:
+    void initializeContext() override
+    {
+        ExecutorTest::initializeContext();
+    }
+};
+
+/// todo add more tests
+TEST_F(SpillAggregationTestRunner, SimpleCase)
+try
+{
+    DB::MockColumnInfoVec column_infos{{"a", TiDB::TP::TypeLongLong}, {"b", TiDB::TP::TypeLongLong}, {"c", TiDB::TP::TypeLongLong}, {"d", TiDB::TP::TypeLongLong}, {"e", TiDB::TP::TypeLongLong}};
+    ColumnsWithTypeAndName column_datas;
+    size_t table_rows = 102400;
+    size_t duplicated_rows = 51200;
+    UInt64 max_block_size = 500;
+    size_t original_max_streams = 20;
+    size_t total_data_size = 0;
+    for (const auto & column_info : mockColumnInfosToTiDBColumnInfos(column_infos))
+    {
+        ColumnGeneratorOpts opts{table_rows, getDataTypeByColumnInfoForComputingLayer(column_info)->getName(), RANDOM, column_info.name};
+        column_datas.push_back(ColumnGenerator::instance().generate(opts));
+        total_data_size += column_datas.back().column->byteSize();
+    }
+    for (auto & column_data : column_datas)
+        column_data.column->assumeMutable()->insertRangeFrom(*column_data.column, 0, duplicated_rows);
+    context.addMockTable("spill_sort_test", "simple_table", column_infos, column_datas, 8);
+
+    auto request = context
+                       .scan("spill_sort_test", "simple_table")
+                       .aggregation({Min(col("c")), Max(col("d")), Count(col("e"))}, {col("a"), col("b")})
+                       .build(context);
+    context.context.setSetting("max_block_size", Field(static_cast<UInt64>(max_block_size)));
+    /// disable spill
+    context.context.setSetting("max_bytes_before_external_group_by", Field(static_cast<UInt64>(0)));
+    auto ref_columns = executeStreams(request, original_max_streams, true);
+    /// enable spill
+    context.context.setSetting("max_bytes_before_external_group_by", Field(static_cast<UInt64>(total_data_size / 200)));
+    context.context.setSetting("group_by_two_level_threshold", Field(static_cast<UInt64>(1)));
+    context.context.setSetting("group_by_two_level_threshold_bytes", Field(static_cast<UInt64>(1)));
+    /// don't use `executeAndAssertColumnsEqual` since it takes too long to run
+    /// test single thread aggregation
+    /// need to enable memory tracker since currently, the memory usage in aggregator is
+    /// calculated by memory tracker, if memory tracker is not enabled, spill will never be triggered.
+    ASSERT_COLUMNS_EQ_UR(ref_columns, executeStreams(request, 1, true));
+    /// test parallel aggregation
+    ASSERT_COLUMNS_EQ_UR(ref_columns, executeStreams(request, original_max_streams, true));
+    /// enable spill and use small max_spilled_size_per_spill
+    context.context.setSetting("max_spilled_size_per_spill", Field(static_cast<UInt64>(total_data_size / 200)));
+    /// test single thread aggregation
+    ASSERT_COLUMNS_EQ_UR(ref_columns, executeStreams(request, 1, true));
+    /// test parallel aggregation
```

Comment:
how about fine grained shuffle?

Reply:
Fine grained shuffle should be the same as `executeStreams` with concurrency 1, as a simple case, I think we don't need to add fine grained case here. We will add more tests later.

Reply:
ok